File: implement-shell-tools/wc/wc.py
import sys
import argparse
import os

parser = argparse.ArgumentParser(
    prog="wc",
    description="Word count",
)

parser.add_argument("-l", action="store_true", help="Count lines")
parser.add_argument("-w", action="store_true", help="Count words")
parser.add_argument("-c", action="store_true", help="Count bytes")
parser.add_argument("paths", nargs="+", help="The files to process")

args = parser.parse_args()

totalLines = 0
totalWords = 0
totalBytes = 0

for path in args.paths:
    try:
        with open(path, "rb") as f:
            buffer = f.read()
        content = buffer.decode("utf-8") #, errors="ignore"
            
        lines = len(content.split("\n")) - 1
        wordCount = len(content.strip().split())
        byte_count = len(buffer)
        
        totalLines += lines
        totalWords += wordCount
        totalBytes += byte_count
        
        if args.l:
            print(f"{lines}      {path}")
        elif args.w:
            print(f"{wordCount}      {path}")
        elif args.c:
            print(f"{bytes}      {path}")
        else:
            print(f"{lines}      {wordCount}       {bytes}      {path}")
    except Exception as e:
        print(f"Error reading {path}: {e}")
        
if len(args.paths) > 1:
    if args.l:
        print(f"{totalLines}      total")
    elif args.w:
        print(f"{totalWords}      total")
    elif args.c:
        print(f"{totalBytes}      total")
    else:
        print(f"{totalLines}      {totalWords}      {totalBytes}      total")

File: implement-shell-tools/wc/wc.js
import { program } from "commander";
import { promises as fs } from "node:fs";

program
    .name("wc")
    .description("Word count")
    .option("-l", "Count lines")
    .option("-w", "Count words")
    .option("-c", "Count bytes")
    .argument("<path...>", "Files to process");

program.parse();

const paths = program.args;
const flags = program.opts();

let totalLines = 0;
let totalWords = 0;
let totalBytes = 0;

for (const path of paths) {
    try {
        const buffer = await fs.readFile(path);
        const content = buffer.toString();

        const lines = content.split("\n").length - 1
        const wordCount = content.trim().split(/\s+/).length
        const bytes = buffer.length;

        totalLines += lines;
        totalWords += wordCount;
        totalBytes += bytes;

        if (flags.l) {
            console.log(`${lines}      ${path}`)
        } else if (flags.w) {
            console.log(`${wordCount}      ${path}`)
        } else if (flags.c) {
            console.log(`${bytes}      ${path}`)
        }
        else {
            console.log(`${lines}      ${wordCount}       ${bytes}      ${path}`)
        }
    } catch (err) {
        console.error(`wc - ${path}: ${err.message}`)
    }
}

if (paths.length > 1) {
    if (flags.l) {
        console.log(`${totalLines}      total`)
    } else if (flags.w) {
        console.log(`${totalWords}      total`)
    } else if (flags.c) {
        console.log(`${totalBytes}      total`)
    } else {
        console.log(`${totalLines}      ${totalWords}      ${totalBytes}      total`)
    }
}
